src/Datadog.Trace.ClrProfiler.Native/clr_helpers.cpp

```diff
@@ -682,4 +683,212 @@ bool TryParseSignatureTypes(const ComPtr<IMetaDataImport2>& metadata_import,
+          // The compiler will never make method calls to generic methods without
+          // the generic context, so we never expect to hit this at
+          // run-time. If we are evaluating the MethodDef/MethodRef of a generic
+          // method return false because it is invalid.
+          if (generic_count > 0) {
+            return false;
+          }
+
+          hr = metadata_import->GetMemberProps(
+              targetFunctionToken, &parent_token, nullptr, 0, nullptr,
+              nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
+              nullptr);
+          if (SUCCEEDED(hr)) {
+            hr = metadata_import->GetTypeSpecFromToken(parent_token, &spec_signature,
+                                                  &spec_signature_length);
+          }
+          break;
+        case mdtMethodSpec:
+          hr = metadata_import->GetMethodSpecProps(targetFunctionToken,
+                                                  &parent_token, &spec_signature, &spec_signature_length);
+          break;
+        default:
+          Warn("[trace::ReturnTypeIsValueTypeOrGeneric] element_type=", ret_type, ": function token was not a MemberRef, MethodDef, or MethodSpec");
+          return false;
+      }
+
+      if (FAILED(hr)) {
+        Warn("[trace::ReturnTypeIsValueTypeOrGeneric] element_type=", ret_type, ": failed to get parent token or signature");
+        return false;
+      }
+
+      // Determine the index of GenArgCount in the signature
+      size_t parent_token_index;
+      if (token_type == mdtMemberRef || token_type == mdtMethodDef) {
+        // TypeSpec Format: GENERICINST (CLASS | VALUETYPE) TypeDefOrRefEncoded GenArgCount Type Type*
+        // Skip over TypeDefOrRefEncoded by parsing the signature at index 2
+        parent_token_index = 2;
+        mdToken dummy_token;
+        ULONG token_length = CorSigUncompressToken(
+            &spec_signature[parent_token_index], &dummy_token);
+        parent_token_index += token_length;
+      } else if (token_type == mdtMethodSpec) {
+        // MethodSpec Format: GENRICINST GenArgCount Type Type*
+        parent_token_index = 1;
+      } else {
+        Warn("[trace::ReturnTypeIsValueTypeOrGeneric] element_type=", ret_type, ": token_type (", token_type , ") not recognized");
+        return false;
+      }
+
+      // Read the value of GenArgCount in the signature
+      ULONG num_generic_arguments;
+      parent_token_index += CorSigUncompressData(
+          &spec_signature[parent_token_index], &num_generic_arguments);
+
+      // Get a pointer to first type after GenArgCount that we can increment to read the signature
+      PCCOR_SIGNATURE p_current_byte = spec_signature + parent_token_index;
+
+      // Iterate to specified generic type argument index and return the appropriate class token or TypeSpec
+      for (size_t i = 0; i < num_generic_arguments; i++) {
+        if (i != generic_type_index) {
+          if (!ParseType(&p_current_byte)) {
```

Comment:
This `ParseType` operation is used to iterate through the generic type arguments in the TypeSpec|MethodSpec signature. This requires basically a fully working parser so we can correctly determine where to find the next type.

Reply:
It would be good to take this method and use it for parsing the whole signature for matching as well. Not necessary for this change, though.

Reply:
Agreed, we can definitely extend it in the future